keymanager/src/crypto/kdf.rs

```diff
@@ -288,60 +331,195 @@ impl Kdf {
-    }
-
-    /// Status of the internal state, i.e. checksum and runtime signing key.
+    /// The state is considered up-to-date if all generations of the master secret are encrypted
+    /// and stored locally, and the checksum of the last generation matches the given checksum.
+    /// If this condition is not met, the internal state is reset and the KDF needs to be
+    /// initialized again.
     ///
-    /// If given checksum and generation don't match internal state,
-    /// the state is reset and an error is returned.
-    pub fn status(
+    /// WARNINGS:
+    /// - Once master secrets have been persisted to disk, it is intended that manual
+    /// intervention by the operator is required to remove/alter them.
+    /// - The first initialization can take a very long time, especially if all generations
+    /// of the master secret must be replicated from other enclaves.
+    #[allow(clippy::too_many_arguments)]
+    pub fn init<M, E>(
         &self,
-        runtime_id: &Namespace,
-        checksum: Vec<u8>,
+        storage: &dyn KeyValue,
+        runtime_id: Namespace,
         generation: u64,
-    ) -> Result<(Vec<u8>, PublicKey)> {
-        let mut inner = self.inner.write().unwrap();
-        inner.verify_runtime_id(runtime_id)?;
+        checksum: Vec<u8>,
+        epoch: EpochTime,
+        master_secret_fetcher: M,
+        ephemeral_secret_fetcher: E,
+    ) -> Result<State>
+    where
+        M: Fn(u64) -> Result<VerifiableSecret>,
+        E: Fn(EpochTime) -> Result<Secret>,
+    {
+        // If the key manager has no secrets, nothing needs to be replicated.
+        if checksum.is_empty() {
+            let mut inner = self.inner.write().unwrap();
+            inner.set_runtime_id(runtime_id)?;
+
+            if inner.checksum.is_some() {
+                inner.reset();
+                return Err(KeyManagerError::StateCorrupted.into());
+            }
 
-        let local_checksum = inner.get_checksum()?;
-        if !checksum.is_empty() && local_checksum != checksum {
-            // The caller provided a checksum and there was a mismatch.
+            return Ok(State {
+                checksum,
+                next_checksum: inner.next_checksum.clone().unwrap_or_default(),
+                signing_key: inner.signing_key,
+                next_signing_key: inner.next_signing_key,
+            });
+        }
+
+        // Fetch internal state.
+        let (mut next_generation, mut curr_checksum) = {
+            let mut inner = self.inner.write().unwrap();
+            inner.set_runtime_id(runtime_id)?;
+
+            let next_generation = inner.get_next_generation();
+            let curr_checksum = inner.checksum.clone().unwrap_or(runtime_id.0.to_vec());
+            (next_generation, curr_checksum)
+        };
+
+        // On startup replicate ephemeral secrets.
+        if next_generation == 0 {
+            let last = epoch + 1;
+            for epoch in (0..=last).rev() {
```

Comment:
This seems unbounded, is this safe? I assume the idea is that sooner or later replication will fail and then we will stop. But if it never fails, we should still limit the number of epochs here.